pyrasterframes/python/tests/PyRasterFramesTests.py

```diff
@@ -245,72 +333,189 @@ def test_exists_for_all(self):
+        # Try to create a tile from numpy.
+        self.assertEqual(Tile(np.random.randn(10, 10), CellType.int8()).dimensions(), [10, 10])
 
-        df = df.withColumn("any_point", some_point_udf(df.poly_geom))
-        # spark-side UDF/UDT are correct
-        intersect_total = df.agg(sum(
-            st_intersects(df.poly_geom, df.any_point).astype('double')
-        ).alias('s')).collect()[0].s
-        self.assertTrue(intersect_total == df.count())
+        tiles = [Tile(np.random.randn(10, 12), CellType.float64()) for _ in range(3)]
+        to_spark = pd.DataFrame({
+            't': tiles,
+            'b': ['a', 'b', 'c'],
+            'c': [1, 2, 4],
+        })
+        rf_maybe = self.spark.createDataFrame(to_spark)
 
-        # Collect to python driver in shapely UDT
-        pandas_df_out = df.toPandas()
+        # rf_maybe.select(rf_render_matrix(rf_maybe.t)).show(truncate=False)
 
-        # Confirm we get a shapely type back from st_* function and UDF
-        self.assertIsInstance(pandas_df_out.poly_geom.iloc[0], shapely.geometry.Polygon)
-        self.assertIsInstance(pandas_df_out.any_point.iloc[0], shapely.geometry.Point)
+        # Try to do something with it.
+        sums = to_spark.t.apply(lambda a: a.cells.sum()).tolist()
+        maybe_sums = rf_maybe.select(rf_tile_sum(rf_maybe.t).alias('tsum'))
+        maybe_sums = [r.tsum for r in maybe_sums.collect()]
+        np.testing.assert_almost_equal(maybe_sums, sums, 12)
 
-        # And our spark-side manipulations were correct
-        xs_correct = pandas_df_out.point_geom.apply(lambda g: g.coords[0][0]) == pandas_df.x
-        self.assertTrue(all(xs_correct))
+        # Test round trip for an array
+        simple_array = Tile(np.array([[1, 2], [3, 4]]), CellType.float64())
+        to_spark_2 = pd.DataFrame({
+            't': [simple_array]
+        })
 
-        centroid_ys = pandas_df_out.poly_geom.apply(lambda g:
-                                                    g.centroid.coords[0][1]).tolist()
-        numpy.testing.assert_almost_equal(centroid_ys, pandas_df.y.tolist())
+        rf_maybe_2 = self.spark.createDataFrame(to_spark_2)
+        #print("RasterFrame `show`:")
+        #rf_maybe_2.select(rf_render_matrix(rf_maybe_2.t).alias('t')).show(truncate=False)
 
-        # Including from UDF's
-        numpy.testing.assert_almost_equal(
-            pandas_df_out.poly_geom.apply(lambda g: g.area).values,
-            pandas_df_out.poly_area.values
-        )
-        numpy.testing.assert_almost_equal(
-            pandas_df_out.poly_geom.apply(lambda g: g.length).values,
-            pandas_df_out.poly_len.values
-        )
+        pd_2 = rf_maybe_2.toPandas()
+        array_back_2 = pd_2.iloc[0].t
+        #print("Array collected from toPandas output\n", array_back_2)
+
+        self.assertIsInstance(array_back_2, Tile)
+        np.testing.assert_equal(array_back_2.cells, simple_array.cells)
+        np.int8
+
+class RasterSource(TestEnvironment):
```

Comment:
The method `test_raster_join` is now inside this class, which should be somewhere else. Perhaps `RasterFunctions` or make a new one?

Reply:
👍